src/expressions/method-call-expr.md

```diff
@@ -14,38 +14,68 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
 # assert!(1.14 < log_pi && log_pi < 1.15)
 ```
 
-When resolving method calls on an expression of type `A`, Rust will use the
-following order, only looking at methods that are [visible]. If the type of `A`
-is a type parameter or `Self` in a trait definitition then steps 2-4 first
-consider traits from bounds on the type paramter, then the traits that are in
-scope. For other types, only the traits that are in scope are considered.
-
-1. Inherent methods, with receiver of type `A`, `&A`, `&mut A`.
-1. Trait methods with receiver of type `A`.
-1. Trait methods with receiver of type `&A`.
-1. Trait methods with receiver of type `&mut A`.
-1. If it's possible, Rust will then repeat steps 1-5 with
-  `<A as std::ops::Deref>::Target`, and insert a dereference operator.
-1. If `A` is now an [array] type, then repeat steps 1-4 with the corresponding
-  slice type.
-
-Note: In steps 1-4, the receiver is used, not the type of `Self` nor the
-type of `A`. For example:
-
-```rust,ignore
-// `Self` is `&A`, receiver is `&A`.
-impl<'a> Trait for &'a A {
-    fn method(self) {}
+When resolving method calls on an expression of type `A`, Rust looks up methods
+both on the type itself and the traits in implements. Additionally, unlike with
+non-method function calls, the `self` parameter is special and may be
+automatically dereferenced in order to resolve it. Rust uses the following
+process to resolve method calls.
+
+First, Rust will attempt to build a list of candidate receiver types. It obtains
+these by repeatedly [dereferencing][dereference] the type, adding each type
+encountered to the list, then finally attempting an [unsized coercion] at the
+end, and adding the result type if that is successful. Then, for each candidate
+`T`, Rust adds `&T` and `&mut T` to the list immediately afterward.
+
+So, for instance, if `A` is `Box<[i32;2]>`, then the candidate types will be
+`Box<[i32;2]>`, `&Box<[i32;2]>`, `&mut Box<[i32;2]>`, `[i32; 2]` (by
+dereferencing), `&[i32; 2]`, `&mut [i32; 2]`, `[i32]` (by unsized coercion),
+`&[i32]`, and finally `&mut [i32]`.
+
+Then, for each candidate type `T`, Rust will search for a [visible] method with
+a receiver of that type in the following places:
+
+1. `T`'s inherent methods (methods implemented directly on `T`).
+1. Any of the methods provided by a trait implemented by `T`. If `T` is
+   a type parameter (including the `Self` parameter of a trait), then only
+   methods from the trait constraints on `T` are available for lookup. If `T` is
+   not, then methods from any in-scope trait are available.
+
+Note that the lookup is done for each type in order, which can occasionally lead
```

Comment:
Colon after "Note". Entire note and example should be in blockquotes. Could raise to a "Warning:" even.